prometheus/promhttp/option.go

```diff
@@ -56,3 +76,24 @@ func WithExemplarFromContext(getExemplarFn func(requestCtx context.Context) prom
 		o.getExemplarFn = getExemplarFn
 	})
 }
+
+// WithLabelResolver register a label for dynamic resolution with access to request information.
+// Values returned by the resolver that would not be present in allowedValues would be replaced
+// with "unknown" value instead. Values in allowedValues are used as is.
+// See the example for ExampleInstrumentHandlerWithLabelResolver for example usage
+func WithLabelResolver(label string, allowedValues []string, resolver LabelResolver) Option {
```

Comment:
I think I like it, but I think perhaps we could embed that rather into each metric implementation. Some options:

1.  Add `SafeWith` and `SafeWithLabelValues` in e.g. `Observer` (not sure if `Safe` is a good word)

```
type ObserverVec interface {
	GetMetricWith(Labels) (Observer, error)
	GetMetricWithLabelValues(lvs ...string) (Observer, error)
	With(Labels) Observer
	WithLabelValues(...string) Observer
        SafeWith(Labels) Observer // <---
	SafeWithLabelValues(...string) Observer // <---
	CurryWith(Labels) (ObserverVec, error)
	MustCurryWith(Labels) ObserverVec

	Collector
}
```

2. Let metric constructor option for adding label values that are "enforced" or "safe"

If we like this approach, I think I am ok in doing this in another PR and adding some dynamic labels options in middleware and tripperware here.

WDYT @kakkoyun?